[quant] Add support for symmetric quant in executorch

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90304

Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant] Add support for symmetric quant in executorch backend

Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cccclai reviewed (approved):
lgtm

jcaip commented on torch/ao/quantization/backend_config/executorch.py:
```diff
@@ -47,6 +47,42 @@
+# TODO: add additional restriction on qscheme to ensure it
+# is either per_tensor_symmetric or per_channel_symmetric
+
+executorch_act_qint8_scale_min_2_neg_12 = DTypeWithConstraints(
```
I think you are missing import here? Also idk if related but DTypeWithConstraints was giving me errors with `test_public_bindings`, i had to add it to the `__all__` in `backend_config/__init__.py` to get it to go away.

jerryzh168 replied:
I see, makes sense, will update

jerryzh168 committed:
Update on "[quant] Add support for symmetric quant in executorch backend"


Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/backend_config/executorch.py:
```diff
@@ -47,6 +53,42 @@
+# TODO: add additional restriction on qscheme to ensure it
+# is either per_tensor_symmetric or per_channel_symmetric
+
+executorch_act_qint8_scale_min_2_neg_12 = DTypeWithConstraints(
```
is executorch a user facing concept in OSS?  If not, is it ok to use the name here?

jerryzh168 replied:
looks like it's not oss'ed yet, this file was added at the time that we didn't give much thought about this. this will be oss'ed and we probably have mentioned executorch in the recent pt conference so I feel it should be OK?

vkuzo commented on torch/ao/quantization/backend_config/executorch.py:
```diff
@@ -47,6 +53,42 @@
+    scale_min_lower_bound=2 ** -12,
+)
+
+executorch_weight_qint8_neg_127_to_127_scale_min_2_neg_12 = DTypeWithConstraints(
```
is there a difference between restrictions for executorch vs restrictions for qnnpack_running_xnnpack?  If yes, what are they?  If not, can we use the same objects?

I thought executorch was not specific to a device/backend, I'd imagine the restrictions like these should be for specific device/backend combination.  For example, xnnpack on edge, or a specific DSP chipset.

jerryzh168 replied:
it's the same, sure, I can import the same object from qnnpack.py

jerryzh168 committed:
Update on "[quant] Add support for symmetric quant in executorch"


Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add support for symmetric quant in executorch"


Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add support for symmetric quant in executorch"


Summary:
This PR adds symmetric quant in the backend config for executorch

Test Plan:
NA, will be tested in meta internal flow

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@vkuzo let's continue discussing the executorch question offline, I'll land this to unblock the workflows first

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge -f "failed ci tests are not related to the PR"
